benchalerts/clients.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2022, Voltron Data.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import os
+from typing import Optional
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+
+from .log import fatal_and_log, log
+
+
+class _BaseClient(abc.ABC):
+    """A client to interact with an API."""
+
+    base_url: str
+    timeout_s = 10
+
+    def __init__(self):
+        retry_strategy = Retry(
+            total=5,
+            status_forcelist=frozenset((429, 502, 503, 504)),
+            backoff_factor=4,  # will retry in 2, 4, 8, 16, 32 seconds
+        )
+        if os.getenv("UNIT_TESTING"):
+            from tests.mocks import MockAdapter
+
+            adapter = MockAdapter()
```

Comment:
Yeah, it seems like everyone has a different way to mock `requests`. I took this strategy from [arrow-benchmarks-ci](https://github.com/voltrondata-labs/arrow-benchmarks-ci/blob/main/integrations/__init__.py#L58) (along with a moderate amount of other code in there). Since outside of tests we're already using a custom adapter to mount to the `Session`, it seems like a nice opportunity to be able to switch that out with a mocked adapter.

I agree that passing it in through `__init__()` might be a little cleaner; I'll try to mess around with it some more.

Reply:
done; take a look